service/matching/version_sets.go

```diff
@@ -0,0 +1,209 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/dgryski/go-farm"
+	"github.com/pborman/uuid"
+	"go.temporal.io/api/serviceerror"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/util"
+)
+
+func ToBuildIdOrderingResponse(g *persistence.VersioningData, maxDepth int) *workflowservice.GetWorkerBuildIdCompatabilityResponse {
+	return depthLimiter(g, maxDepth)
+}
+
+// HashVersioningData returns a farm.Fingerprint64 hash of the versioning data as bytes. If the data is nonexistent or
+// invalid, returns nil.
+func HashVersioningData(data *persistence.VersioningData) []byte {
+	if data == nil || data.GetVersionSets() == nil {
+		return nil
+	}
+	asBytes, err := data.Marshal()
+	if err != nil {
+		return nil
+	}
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, farm.Fingerprint64(asBytes))
+	return b
+}
+
+func depthLimiter(g *persistence.VersioningData, maxDepth int) *workflowservice.GetWorkerBuildIdCompatabilityResponse {
+	if maxDepth <= 0 || maxDepth >= len(g.GetVersionSets()) {
+		return &workflowservice.GetWorkerBuildIdCompatabilityResponse{MajorVersionSets: g.GetVersionSets()}
+	}
+	shortened := util.SliceTail(g.GetVersionSets(), maxDepth)
+	return &workflowservice.GetWorkerBuildIdCompatabilityResponse{MajorVersionSets: shortened}
```

Comment:
```suggestion
  sets := g.GetVersionSets()
	if maxDepth > 0 {
	  sets = util.SliceTail(sets, maxDepth)
	}
	return &workflowservice.GetWorkerBuildIdCompatabilityResponse{MajorVersionSets: sets}
```
maybe flip it to avoid an early return and simplify?